src/Language/PureScript/Kinds.hs

```diff
@@ -48,67 +47,65 @@ srcFunKind = FunKind NullSourceAnn
 srcNamedKind :: Qualified (ProperName 'KindName) -> SourceKind
 srcNamedKind = NamedKind NullSourceAnn
 
-instance A.ToJSON a => A.ToJSON (Kind a) where
-  toJSON kind = case kind of
+kindToJSON :: forall a. (a -> Value) -> Kind a -> Value
+kindToJSON annToJSON kind =
+  case kind of
     KUnknown a i ->
-      obj "KUnknown" a i
+      variant "KUnknown" a i
     Row a k ->
-      obj "Row" a k
+      variant "Row" a (go k)
     FunKind a k1 k2 ->
-      obj "FunKind" a [k1, k2]
+      variant "FunKind" a (go k1, go k2)
     NamedKind a n ->
-      obj "NamedKind" a n
-    where
-    obj :: A.ToJSON b => Text -> a -> b -> A.Value
-    obj tag ann contents =
-      A.object [ "tag" .= tag, "annotation" .= ann, "contents" .= contents ]
-
--- This handles JSON generated by compilers up to 0.10.3 and maps them to the
--- new representations (i.e. NamedKinds which are defined in the Prim module).
-kindFromJSON :: Parse Text a -> Parse Text a -> Parse Text (Kind a)
-kindFromJSON defaultAnn annFromJSON = fix $ \go -> do
-  t <- key "tag" asText
-  let annFromJSON' = key "annotation" annFromJSON <|> defaultAnn
-  case t of
+      variant "NamedKind" a n
+  where
+  go :: Kind a -> Value
+  go = kindToJSON annToJSON
+
+  variant :: A.ToJSON b => Text -> a -> b -> A.Value
+  variant tag ann contents =
+    A.object
+      [ "tag"        .= tag
+      , "annotation" .= annToJSON ann
+      , "contents"   .= contents
+      ]
+
+instance A.ToJSON a => A.ToJSON (Kind a) where
+  toJSON = kindToJSON toJSON
+
+kindFromJSON :: forall a. Parser a -> (Value -> Parser a) -> Value -> Parser (Kind a)
+kindFromJSON defaultAnn annFromJSON = A.withObject "Kind" $ \o -> do
+  tag <- o .: "tag"
+  a   <- (o .: "annotation" >>= annFromJSON) <|> defaultAnn
+  let
+    contents :: A.FromJSON b => Parser b
+    contents = o .: "contents"
+  case tag of
     "KUnknown" ->
-      KUnknown <$> annFromJSON' <*> key "contents" (nth 0 asIntegral)
-    "Star" ->
-      kindType <$> defaultAnn
+      KUnknown a <$> contents
```

Comment:
I'm pretty sure this parser was broken. It appears to parse with `nth 0`, but it was not encoded with a wrapper list.